HopsanCore/src/ComponentUtilities/TempDirectoryHandle.cpp

```diff
@@ -0,0 +1,263 @@
+/*-----------------------------------------------------------------------------
+
+ Copyright 2017 Hopsan Group
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+
+ The full license is available in the file LICENSE.
+ For details about the 'Hopsan Group' or information about Authors and
+ Contributors see the HOPSANGROUP and AUTHORS files that are located in
+ the Hopsan source code root directory.
+
+-----------------------------------------------------------------------------*/
+
+//!
+//! @file   TempDirectoryHandle.cpp
+//! @author Robert Braun <robert.braun@liu.se>
+//! @date   2021-02-02
+//!
+//! @brief Contains a temp directory handle class
+//!
+//$Id$
+
+#if true //__cplusplus < 201103L
+#include <stdlib.h>
+#include "ComponentUtilities/num2string.hpp"
+#endif
+
+#include <iostream>
+#if __cplusplus >= 201703L
+#include <filesystem>
+#endif
+#include <random>
+#include <sstream>
+#include <iomanip>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <cstring>
+#include <stdio.h>
+#include <stdlib.h>
+#ifdef _WIN32
+#include <windows.h>
+#include <direct.h>
+#include <shellapi.h>
+#include <strsafe.h>
+#else
+#include <dirent.h>
+#include <unistd.h>
+#endif
+
+
+
+#include "ComponentUtilities/TempDirectoryHandle.h"
+
+using namespace hopsan;
+
+TempDirectoryHandle::TempDirectoryHandle(HString prefix)
+{
+        mPath = getTempDirectory();
+        mPath.append("/Hopsan/"+prefix);
+        mPath.append(generateRandomNumericString());
+
+#if __cplusplus >= 201703L
+        mPath = std::filesystem::create_directory(mPath.c_str());
+#else
+        mIsValid = createDirectory(mPath);
+#endif
+}
+
+TempDirectoryHandle::~TempDirectoryHandle()
+{
+    std::cout << "Test\n";
+    removeDirectory(mPath);
+}
+
+HString TempDirectoryHandle::path()
+{
+    return mPath;
+}
+
+bool TempDirectoryHandle::isValid()
+{
+    return mIsValid;
+}
+
+
+HString TempDirectoryHandle::getTempDirectory() {
+#if __cplusplus >= 201703L
+        mPath = std::filesystem::temp_directory_path().c_str();
+#else
+#ifdef _WIN32
+    char tempdirbuff[MAX_PATH+1];
+    GetTempPathA(MAX_PATH+1, tempdirbuff);
+    return HString(tempdirbuff);
```

Comment:
should this set mPath and mIsValid ?

Reply:
I changed to `path`, as was intended.

Reply:
this still looks odd, the c++17 case assigns mPath, and does not return (from a function returning HString)
In thr other case it returns a Hstring, Also return HString should not be const, since it returns by value

Reply:
I changed it to a return value. I guess I didn't notice, since you have to manually turn on C++17 compilation.